### PR TITLE
Fixed `num_sanity_val_steps` affecting reproducibility of training data shuffling

### LINKED ISSUE #6988: Val sanity steps affect seed reproducibility of train shuffling
## 🐛 Bug

Despite using a separate val dataset, the default sanity val steps affect the indices returned by the train dataloader when shuffling is enabled. The training still executes fine, but this makes it impossible to replicate the exact results of a vanilla PyTorch training (which likely doesn't do a sanity val check) using the same random seed and default trainer settings.

Currently, you can work around this by setting `trainer.num_sanity_val_steps=0` but I think it's bad practice to encourage that. Or you might not care about reproducing exact results, however I find that is a crucial step in refactoring from vanilla PyTorch to Lightning. Otherwise, you don't know whether it's your random seeds causing a performance difference or something more serious with the optimizer(s), loss, scheduler(s), data loaders, etc.

### To Reproduce

Run this Colab [notebook](https://colab.research.google.com/drive/1kYLA_BFLZa9eL7D2EbD_BQ502EvW0Vx7?usp=sharing
) with the `BoringModel`

### Expected behavior

The shuffling of train batches should not be affected by whether we've run sanity steps with the val data loader

### Environment

Automated output provided by Colab notebook

* CUDA:
	- GPU: Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.18.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0+cu101
	- pytorch-lightning: 0.10.0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:  64bit
	- processor:         x86_64
	- python:            3.6.9
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### Additional context

This is my first time working with Colab, so let me know if there are any permissions or other issues that need to be resolved

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #6988

**Test fails on master**

**Important note**
This fix will impact reproducibility from one version to the next as the trainer runs a number of sanity validation steps by default.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -581,6 +583,8 @@ def run_train(self) -> None:
 
         self.checkpoint_connector.has_trained = False
 
+        reset_seed()
+
```
Comment (awaelchli) [inline pytorch_lightning/trainer/trainer.py:587]: adding this here is the fix

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 2]
```diff
@@ -180,9 +180,7 @@ def _call_children_scripts(self):
 
     def setup_distributed(self):
         # TODO: check if needed
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp.py:182]:
```suggestion
```

File: pytorch_lightning/plugins/training_type/ddp_spawn.py [hunk 3]
```diff
@@ -129,9 +129,7 @@ def new_process(self, process_idx, trainer, mp_queue):
         self.mp_queue = mp_queue
 
         # TODO: check if needed
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/ddp_spawn.py:131]:
```suggestion
```

File: tests/checkpointing/test_trainer_checkpoint.py [hunk 4]
```diff
@@ -68,6 +69,7 @@ def validation_step(self, batch, batch_idx):
             limit_train_batches=12,
             limit_val_batches=12,
             limit_test_batches=12,
+            num_sanity_val_steps=0,
```
Comment (awaelchli) [inline tests/checkpointing/test_trainer_checkpoint.py:72]: not sure yet what's going on there (why this fixes the test)

File: tests/callbacks/test_pruning.py [hunk 5]
```diff
@@ -250,16 +250,18 @@ def test_multiple_pruning_callbacks(tmpdir, caplog, make_pruning_permanent: bool
 
     actual = [m.strip() for m in caplog.messages]
     actual = [m for m in actual if m.startswith("Applied")]
+    from pprint import pprint
+    pprint(actual)
```
Comment (carmocca) [inline tests/callbacks/test_pruning.py:254]:
```suggestion
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:600]: Can we move this into `run_sanity_check` instead ? It would be cleaner. Mind adding a comment too ?

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:600]: > Can we move this into `run_sanity_check` instead ? It would be cleaner. Mind adding a comment too ?

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: good call. changed!

**Comment by awaelchli** [comment]
windows problems and macOS timing out :)

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/utilities/test_seed.py [hunk 6]
```diff
@@ -53,3 +54,24 @@ def test_out_of_bounds_seed(seed):
...
+    assert "PL_GLOBAL_SEED" not in os.environ
+
+
+def test_reset_seed_everything():
+    """ Test that we can reset the seed to the initial value set by seed_everything() """
+    assert "PL_GLOBAL_SEED" not in os.environ
+    seed_utils.seed_everything(123)
+    assert os.environ["PL_GLOBAL_SEED"] == "123"
+    before = torch.rand(1)
+    _ = torch.rand(1)
```
Comment (kaushikb11) [inline tests/utilities/test_seed.py:74]: Why is this line required?

**Review by kaushikb11 — APPROVED** [review]
LGTM! But there seems to be a failing test

File: tests/utilities/test_seed.py (hunk 6 above)
Comment (awaelchli) [reply tests/utilities/test_seed.py:74]: this line is not required. it was a test for the good reviewers with the eagle eye!

**Review by ananthsub — APPROVED** [review]
